File: src/shared/pricing.js
export const PRICING_VERSION = "openrouter-cache";
export const FALLBACK_PRICE_MAP = {};

export function estimateUsageCost(item, priceMap = {}) {
  const match = resolvePriceKey(item.model, priceMap);
  if (!match) {
    return {
      inputCostUsd: null,
      outputCostUsd: null,
      cacheReadCostUsd: null,
      cacheWriteCostUsd: null,
      reasoningCostUsd: null,
      estimatedCostUsd: null,
      costQuality: "unknown_price",
      pricingVersion: PRICING_VERSION,
      pricingModel: "",
      pricingSource: ""
    };
  }
  const price = priceMap[match.key];
  const cacheReadTokens = Number(item.cacheReadTokens || 0);
  const cacheWriteTokens = Number(item.cacheWriteTokens || 0);
  // inputTokens is the normalized non-cache input fact. Cache remains priced separately.
  const input = tokenCost(item.inputTokens, price.input_cost_per_token);
  const output = tokenCost(item.outputTokens, price.output_cost_per_token);
  const cacheRead = tokenCost(cacheReadTokens, price.cache_read_input_token_cost ?? price.input_cost_per_token);
  const cacheWrite = tokenCost(cacheWriteTokens, price.cache_creation_input_token_cost ?? price.input_cost_per_token);
  const reasoning = tokenCost(item.reasoningTokens, price.reasoning_cost_per_token);
  return {
    inputCostUsd: roundUsd(input),
    outputCostUsd: roundUsd(output),
    cacheReadCostUsd: roundUsd(cacheRead),
    cacheWriteCostUsd: roundUsd(cacheWrite),
    reasoningCostUsd: roundUsd(reasoning),
    estimatedCostUsd: roundUsd(input + output + cacheRead + cacheWrite + reasoning),
    costQuality: match.exact ? "exact_price" : "estimated_price",
    pricingVersion: price.pricingVersion || PRICING_VERSION,
    pricingModel: match.key,
    pricingSource: price.source || ""
  };
}

export function addCostToUsageItem(item, priceMap = {}) {
  return { ...item, ...estimateUsageCost(item, priceMap) };
}

export function aggregateCost(target, source) {
  if (source.estimatedCostUsd === null || source.estimatedCostUsd === undefined) {
    target.missingPriceModels ||= {};
    const model = source.model || "unknown";
    target.missingPriceModels[model] = (target.missingPriceModels[model] || 0) + (source.totalTokens || 0);
    target.missingPriceTokens = (target.missingPriceTokens || 0) + (source.totalTokens || 0);
    target.costQuality = mergeCostQuality(target.costQuality, "unknown_price");
    return;
  }
  target.hasKnownPrice = true;
  target.inputCostUsd = roundUsd((target.inputCostUsd || 0) + Number(source.inputCostUsd || 0));
  target.outputCostUsd = roundUsd((target.outputCostUsd || 0) + Number(source.outputCostUsd || 0));
  target.cacheReadCostUsd = roundUsd((target.cacheReadCostUsd || 0) + Number(source.cacheReadCostUsd || 0));
  target.cacheWriteCostUsd = roundUsd((target.cacheWriteCostUsd || 0) + Number(source.cacheWriteCostUsd || 0));
  target.reasoningCostUsd = roundUsd((target.reasoningCostUsd || 0) + Number(source.reasoningCostUsd || 0));
  target.estimatedCostUsd = roundUsd((target.estimatedCostUsd || 0) + source.estimatedCostUsd);
  target.costQuality = mergeCostQuality(target.costQuality, source.costQuality || "unknown_price");
  target.pricingVersion ||= source.pricingVersion || PRICING_VERSION;
  target.pricingSource ||= source.pricingSource || "";
}

export function mergeCostQuality(left = "", right = "") {
  const rank = { exact_price: 0, estimated_price: 1, unknown_price: 2 };
  if (!left) return right || "";
  if (!right) return left;
  return rank[right] > rank[left] ? right : left;
}

export function createPriceMap(customPrices = {}, remotePrices = {}) {
  const remoteMap = {};
  for (const price of Object.values(remotePrices || {})) {
    const key = normalizeModelName(price.model || price.key || "");
    if (!key) continue;
    remoteMap[key] = normalizePriceRecord(price, {
      source: price.source || "openrouter",
      pricingVersion: price.pricingVersion || PRICING_VERSION
    });
  }
  const customMap = {};
  for (const price of Object.values(customPrices || {})) {
    const key = normalizeModelName(price.model || price.key || "");
    if (!key) continue;
    customMap[key] = normalizePriceRecord(price, {
      source: price.source || "custom",
      pricingVersion: "custom-overrides"
    });
  }
  return { ...remoteMap, ...customMap };
}

export function normalizeModelName(model) {
  return String(model || "")
    .trim()
    .toLowerCase()
    .replace(/^~/, "")
    .replace(/^openrouter\//, "");
}

export function priceToPublic(model, price) {
  return {
    model,
    inputCostPerMTok: toPerMillion(price.input_cost_per_token),
    outputCostPerMTok: toPerMillion(price.output_cost_per_token),
    cacheReadCostPerMTok: toPerMillion(price.cache_read_input_token_cost),
    cacheWriteCostPerMTok: toPerMillion(price.cache_creation_input_token_cost),
    reasoningCostPerMTok: 0,
    source: price.source || "builtin",
    pricingVersion: price.pricingVersion || "",
    updatedAt: price.updatedAt || ""
  };
}

export function openRouterModelToPrice(model, fetchedAt = "", pricingVersion = PRICING_VERSION) {
  const pricing = model?.pricing || {};
  const prompt = numberOrNull(pricing.prompt);
  const completion = numberOrNull(pricing.completion);
  if (prompt === null && completion === null) return null;
  return {
    model: normalizeModelName(model.id || model.canonical_slug || ""),
    displayName: model.name || "",
    canonicalSlug: model.canonical_slug || "",
    input_cost_per_token: prompt || 0,
    output_cost_per_token: completion || 0,
    cache_read_input_token_cost: numberOrNull(pricing.input_cache_read) ?? prompt ?? 0,
    cache_creation_input_token_cost: numberOrNull(pricing.input_cache_write) ?? prompt ?? 0,
    reasoning_cost_per_token: 0,
    max_input_tokens: Number(model.context_length || model.top_provider?.context_length || 0),
    max_output_tokens: Number(model.top_provider?.max_completion_tokens || 0),
    source: "openrouter",
    pricingVersion,
    updatedAt: fetchedAt
  };
}

function resolvePriceKey(model = "", priceMap = {}) {
  const normalized = normalizeModelName(model);
  if (!normalized) return null;
  if (priceMap[normalized]) return { key: normalized, exact: true };
  for (const candidate of createMatchingCandidates(normalized)) {
    if (priceMap[candidate]) return { key: candidate, exact: false };
  }
  for (const key of Object.keys(priceMap)) {
    if (key.includes(normalized) || normalized.includes(key)) return { key, exact: false };
  }
  return null;
}

function createMatchingCandidates(model) {
  const withoutProvider = model.includes("/") ? model.split("/").at(-1) : model;
  return [
    withoutProvider,
    `openai/${model}`,
    `anthropic/${model}`,
    `google/${model}`,
    `meta-llama/${model}`,
    `qwen/${model}`,
    `deepseek/${model}`,
    `mistralai/${model}`,
    `openai/${withoutProvider}`,
    `anthropic/${withoutProvider}`,
    `google/${withoutProvider}`,
    `meta-llama/${withoutProvider}`,
    `qwen/${withoutProvider}`,
    `deepseek/${withoutProvider}`,
    `mistralai/${withoutProvider}`
  ];
}

function normalizePriceRecord(price, defaults = {}) {
  return {
    input_cost_per_token: toPerToken(price.inputCostPerMTok ?? price.inputCostPerToken ?? price.input_cost_per_token, "inputCostPerMTok" in price),
    output_cost_per_token: toPerToken(price.outputCostPerMTok ?? price.outputCostPerToken ?? price.output_cost_per_token, "outputCostPerMTok" in price),
    cache_read_input_token_cost: toPerToken(price.cacheReadCostPerMTok ?? price.cacheReadCostPerToken ?? price.cache_read_input_token_cost, "cacheReadCostPerMTok" in price),
    cache_creation_input_token_cost: toPerToken(price.cacheWriteCostPerMTok ?? price.cacheWriteCostPerToken ?? price.cache_creation_input_token_cost, "cacheWriteCostPerMTok" in price),
    reasoning_cost_per_token: 0,
    source: price.source || defaults.source || "custom",
    pricingVersion: price.pricingVersion || defaults.pricingVersion || "",
    updatedAt: price.updatedAt || ""
  };
}

function tokenCost(tokens, costPerToken = 0) {
  const n = Number(tokens || 0);
  const price = Number(costPerToken || 0);
  return Number.isFinite(n) && Number.isFinite(price) ? n * price : 0;
}

function roundUsd(value) {
  return Math.round(Number(value || 0) * 1_000_000) / 1_000_000;
}

function toPerToken(value, isPerMillion = false) {
  const n = Number(value || 0);
  if (!Number.isFinite(n) || n < 0) return 0;
  return isPerMillion ? n / 1_000_000 : n;
}

function toPerMillion(value) {
  return Math.round(Number(value || 0) * 1_000_000 * 1_000_000) / 1_000_000;
}

function numberOrNull(value) {
  const n = Number(value);
  return Number.isFinite(n) && n >= 0 ? n : null;
}

File: collector-core/src/pricing.rs
use serde_json::Value;
use std::collections::HashMap;

pub const PRICING_VERSION: &str = "openrouter-cache";

#[derive(Debug, Clone)]
pub struct CostResult {
    pub input_cost_usd: Option<f64>,
    pub output_cost_usd: Option<f64>,
    pub cache_read_cost_usd: Option<f64>,
    pub cache_write_cost_usd: Option<f64>,
    pub reasoning_cost_usd: Option<f64>,
    pub estimated_cost_usd: Option<f64>,
    pub cost_quality: String,
    pub pricing_version: String,
    pub pricing_model: String,
    pub pricing_source: String,
}

fn round_usd(v: f64) -> f64 {
    (v * 1_000_000.0).round() / 1_000_000.0
}

fn token_cost(tokens: i64, cost_per_token: f64) -> f64 {
    tokens as f64 * cost_per_token
}

fn normalize_model_name(model: &str) -> String {
    let mut s = model.trim().to_lowercase();
    if s.starts_with('~') {
        s = s[1..].to_string();
    }
    if let Some(rest) = s.strip_prefix("openrouter/") {
        s = rest.to_string();
    }
    s
}

fn resolve_price_key<'a>(model: &str, price_map: &'a HashMap<String, Value>) -> Option<(String, bool, &'a Value)> {
    let normalized = normalize_model_name(model);
    if let Some(v) = price_map.get(&normalized) {
        return Some((normalized, true, v));
    }
    // Try fuzzy matching candidates
    let candidates = create_matching_candidates(&normalized);
    for candidate in candidates {
        if let Some(v) = price_map.get(&candidate) {
            return Some((candidate, false, v));
        }
    }
    None
}

fn create_matching_candidates(normalized: &str) -> Vec<String> {
    let mut candidates = Vec::new();
    let providers = ["openai", "anthropic", "google", "meta-llama", "qwen", "deepseek", "mistralai"];

    // Strip provider prefix if present
    for &provider in &providers {
        if let Some(rest) = normalized.strip_prefix(&format!("{}/", provider)) {
            candidates.push(rest.to_string());
            candidates.push(format!("{}/{}", provider, rest));
        }
    }

    candidates
}

pub fn estimate_usage_cost(item: &Value, price_map: &HashMap<String, Value>) -> CostResult {
    let model = item["model"].as_str().unwrap_or("");

    match resolve_price_key(model, price_map) {
        Some((_key, exact, price)) => {
            let input_per = price["input_cost_per_token"].as_f64().unwrap_or(0.0);
            let output_per = price["output_cost_per_token"].as_f64().unwrap_or(0.0);
            let cache_read_per = price["cache_read_input_token_cost"]
                .as_f64()
                .unwrap_or(input_per);
            let cache_write_per = price["cache_creation_input_token_cost"]
                .as_f64()
                .unwrap_or(input_per);
            let reasoning_per = price["reasoning_cost_per_token"].as_f64().unwrap_or(0.0);

            let input_tokens = item["inputTokens"].as_i64().unwrap_or(0);
            let output_tokens = item["outputTokens"].as_i64().unwrap_or(0);
            let cache_read_tokens = item["cacheReadTokens"].as_i64().unwrap_or(0);
            let cache_write_tokens = item["cacheWriteTokens"].as_i64().unwrap_or(0);
            let reasoning_tokens = item["reasoningTokens"].as_i64().unwrap_or(0);

            let input = token_cost(input_tokens, input_per);
            let output = token_cost(output_tokens, output_per);
            let cache_read = token_cost(cache_read_tokens, cache_read_per);
            let cache_write = token_cost(cache_write_tokens, cache_write_per);
            let reasoning = token_cost(reasoning_tokens, reasoning_per);
            let total = input + output + cache_read + cache_write + reasoning;

            CostResult {
                input_cost_usd: Some(round_usd(input)),
                output_cost_usd: Some(round_usd(output)),
                cache_read_cost_usd: Some(round_usd(cache_read)),
                cache_write_cost_usd: Some(round_usd(cache_write)),
                reasoning_cost_usd: Some(round_usd(reasoning)),
                estimated_cost_usd: Some(round_usd(total)),
                cost_quality: if exact { "exact_price" } else { "estimated_price" }.to_string(),
                pricing_version: price["pricingVersion"].as_str().unwrap_or(PRICING_VERSION).to_string(),
                pricing_model: price["aliasTarget"].as_str().unwrap_or(model).to_string(),
                pricing_source: price["source"].as_str().unwrap_or("").to_string(),
            }
        }
        None => CostResult {
            input_cost_usd: None,
            output_cost_usd: None,
            cache_read_cost_usd: None,
            cache_write_cost_usd: None,
            reasoning_cost_usd: None,
            estimated_cost_usd: None,
            cost_quality: "unknown_price".to_string(),
            pricing_version: PRICING_VERSION.to_string(),
            pricing_model: String::new(),
            pricing_source: String::new(),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_normalize_model_name() {
        assert_eq!(normalize_model_name("GPT-4o"), "gpt-4o");
        assert_eq!(normalize_model_name("~claude-3"), "claude-3");
        assert_eq!(normalize_model_name("openrouter/gpt-4"), "gpt-4");
    }

    #[test]
    fn test_estimate_unknown_price() {
        let item = json!({"model": "unknown-model"});
        let map = HashMap::new();
        let cost = estimate_usage_cost(&item, &map);
        assert_eq!(cost.cost_quality, "unknown_price");
        assert!(cost.estimated_cost_usd.is_none());
    }
}
